packages/http/src/validator/validators/body.ts

```diff
@@ -1,82 +1,131 @@
+import * as Array from 'fp-ts/lib/Array';
+import * as Either from 'fp-ts/lib/Either';
+import * as Option from 'fp-ts/lib/Option';
+import { pipe } from 'fp-ts/lib/pipeable';
+import { get, partial } from 'lodash';
 import * as typeIs from 'type-is';
-import { body } from '../deserializers';
 
-import { DiagnosticSeverity, Dictionary } from '@stoplight/types';
 import { JSONSchema } from '../../types';
-import { validateAgainstSchema } from '../validators/utils';
+import { body } from '../deserializers';
 import { IHttpValidator } from './types';
+import { validateAgainstSchema } from './utils';
 
 export class HttpBodyValidator implements IHttpValidator<any, IMediaTypeContent> {
-  constructor(private _prefix: string) {}
+  constructor(private prefix: string) {}
 
   public validate(target: any, specs: IMediaTypeContent[], mediaType?: string): IPrismDiagnostic[] {
-    const { _prefix: prefix } = this;
-    const content = this.getContent(specs, mediaType);
-
+    const content = getContent(specs, mediaType);
     const schema = get(content, 'schema');
 
-    if (!schema) {
-      return [];
-    }
-
-    if (mediaType && typeIs.is(mediaType, 'application/x-www-form-urlencoded')) {
-      const shallowParsedTarget = shallowParseUrlEncoded(target);
-      target = deepParseUrlEncoded(target);
-
-      const encodings = get(content, 'encodings', []);
-      for (const encoding of encodings) {
-        const allowReserved = get(encoding, 'allowReserved', false);
-        const property = encoding.property;
-        const value = shallowParsedTarget[property];
-
-        if (!allowReserved && typeof value === 'string' && value.match(/[\/?#\[\]@!$&'()*+,;=]/)) {
-          return [
-            {
-              path: [prefix, property],
-              message: 'Reserved characters used in request body',
-              severity: DiagnosticSeverity.Error,
-            },
-          ];
-        }
+    return pipe(
+      schema ? Either.right(target) : Either.left([]),
+      Either.chain(partial(maybeValidateFormBody, schema!, content, mediaType)),
+      Either.chain(partial(validateBody, schema!)),
+      Either.fold(partial(applyPrefix, this.prefix), () => []),
+    );
+  }
+}
 
-        if (encoding.style) {
-          const deserializer = body.get(encoding.style);
-          if (deserializer && schema.properties) {
-            const propertySchema = schema.properties[property];
-            target[property] = deserializer.deserialize(property, target, propertySchema as JSONSchema);
-          }
-        }
-      }
-    }
+function validateBody(schema: JSONSchema, target: any) {
+  const diagnostics = validateAgainstSchema(target, schema);
+  return diagnostics.length ? Either.left(diagnostics) : Either.right(target);
+}
+
+function maybeValidateFormBody(
+  schema: JSONSchema,
+  content: IMediaTypeContent,
+  mediaType: string | undefined,
+  target: any,
+) {
```

Comment:
In general I really suggest to annotate the return type explicitly. This makes the debugger way easier sometimes.

Reply:
Done 👍